lib/connection_tester.rb

```diff
@@ -0,0 +1,243 @@
+
+class ConnectionTester
+  REQUEST_OPTS = {
+    open_timeout: 6,
+    ssl_timeout:  6,
+    read_timeout: 12
+  }
+
+  NODEINFO_SCHEMA   = "http://nodeinfo.diaspora.software/ns/schema/1.0"
+  NODEINFO_FRAGMENT = "/.well-known/nodeinfo"
+
+  class << self
+    # Test the reachability of a server by the given HTTP/S URL.
+    # In the first step, a DNS query is performed to check whether the
+    # given name even resolves correctly.
+    # The second step is to send a HTTP request and look at the returned
+    # status code or any returned errors.
+    # This function isn't intended to check for the availability of a
+    # specific page, instead a GET request is sent to the root directory
+    # of the server.
+    # In the third step an attempt is made to determine the software version
+    # used on the server, via the nodeinfo page.
+    #
+    # @param [String] server URL
+    # @return [Result] result object containing information about the
+    #                  server and to what point the connection was successful
+    def check(url)
+      url = "http://#{url}" unless url.include?('://')
+      res = Result.new(host)
+
+      begin
+        req_opts = request_opts_for_url(url)
+        host = req_opts.delete(:host)
+        ct = ConnectionTester.new(host)
+
+        # test DNS resolving
+        ct.resolve(res)
+
+        # test HTTP request
+        ct.request(req_opts, res)
+
+        # test for the diaspora* version
+        ct.nodeinfo(res)
+
+      rescue Failure => e
+        case e
+        when AddressFailure, DNSFailure, NetFailure
+          res.reachable = false
+        when SSLFailure
+          res.reachable = true
+          res.ssl_status = false
+        when HTTPFailure
+          res.reachable = true
+          res.ssl_status = true if ct.uses_ssl?
+        when NodeInfoFailure
+          res.diaspora_version = ""
+        end
+
+        res.error = e
+      end
+
+      res.freeze
+    end
+
+    # assemble a hash of options for a Net::HTTP request.
+    # This also includes a :host key containing the hostname of the server
+    #
+    # @param [String] target URL
+    def request_opts_for_url(url)
+      uri = URI.parse(url)
+      raise AddressFailure, "invalid protocol: '#{uri.scheme.upcase}'" unless uri.is_a?(URI::HTTP) || uri.is_a?(URI::HTTPS)
+      ssl = (uri.scheme == 'https') ? true : false
```

Comment:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.